File: Android/scenes/ktv/src/main/java/io/agora/scene/ktv/live/KTVApi.kt
package io.agora.scene.ktv.live

import io.agora.mediaplayer.Constants
import io.agora.mediaplayer.IMediaPlayer
import io.agora.musiccontentcenter.IAgoraMusicContentCenter
import io.agora.musiccontentcenter.Music
import io.agora.musiccontentcenter.MusicChartInfo
import io.agora.rtc2.RtcEngine

/**
 * 在KTVApi中的身份
 * @param SoloSinger 独唱者: 当前只有自己在唱歌
 * @param CoSinger 合唱者: 加入合唱需要通过调用switchSingerRole将切换身份成合唱
 * @param LeadSinger 主唱: 有合唱者加入后，需要通过调用switchSingerRole切换身份成主唱
 * @param Audience 观众: 默认状态
 */
enum class KTVSingRole(val value: Int) {
    SoloSinger(0),
    CoSinger(1),
    LeadSinger(2),
    Audience(3)
}

/**
 * loadSong失败的原因
 * @param NO_LYRIC_URL 没有歌词，不影响音乐正常播放
 * @param MUSIC_PRELOAD_FAIL 音乐加载失败
 * @param CANCELED 本次加载已终止
 */
enum class KTVLoadSongFailReason(val value: Int) {
    NO_LYRIC_URL(0),
    MUSIC_PRELOAD_FAIL(1),
    CANCELED(2)
}

/**
 * switchSingerRole的失败的原因
 * @param JOIN_CHANNEL_FAIL 加入channel2失败
 * @param NO_PERMISSION switchSingerRole传入了错误的目标角色（不能从当前角色切换到目标角色）
 */
enum class SwitchRoleFailReason(val value: Int) {
    JOIN_CHANNEL_FAIL(0),
    NO_PERMISSION(1)
}

/**
 * 加载音乐的模式
 * @param LOAD_MUSIC_ONLY 只加载音乐（通常加入合唱前使用此模式）
 * @param LOAD_LRC_ONLY 只加载歌词（通常歌曲开始播放时观众使用此模式）
 * @param LOAD_MUSIC_AND_LRC 默认模式，加载歌词和音乐（通常歌曲开始播放时主唱使用此模式）
 */
enum class KTVLoadMusicMode(val value: Int) {
    LOAD_MUSIC_ONLY(0),
    LOAD_LRC_ONLY(1),
    LOAD_MUSIC_AND_LRC(2)
}

/**
 * 歌词组件接口，您setLrcView传入的歌词组件需要继承此接口类，并实现以下三个方法
 */
interface ILrcView {
    /**
     * ktvApi内部更新音高pitch时会主动调用此方法将pitch值传给你的歌词组件
     * @param pitch 音高值
     */
    fun onUpdatePitch(pitch: Float?)

    /**
     * ktvApi内部更新音乐播放进度progress时会主动调用此方法将进度值progress传给你的歌词组件，50ms回调一次
     * @param progress 歌曲播放的真实进度 50ms回调一次
     */
    fun onUpdateProgress(progress: Long?)

    /**
     * ktvApi获取到歌词地址时会主动调用此方法将歌词地址url传给你的歌词组件，您需要在这个回调内完成歌词的下载
     */
    fun onDownloadLrcData(url: String?)
}

enum class MusicLoadStatus(val value: Int) {
    COMPLETED(0),
    FAILED(1),
    INPROGRESS(2),
}

/**
 * 音乐加载状态接口
 */
interface OnMusicLoadStateListener {
    /**
     * 音乐加载成功
     * @param songCode 歌曲编码， 和你loadMusic传入的songCode一致
     * @param lyricUrl 歌词地址
     */
    fun onMusicLoadSuccess(songCode: Long, lyricUrl: String)

    /**
     * 音乐加载失败
     * @param reason 歌曲加载失败的原因
     */
    fun onMusicLoadFail(songCode: Long, reason: KTVLoadSongFailReason)

    /**
     * 音乐加载进度
     * @param songCode 歌曲编码
     * @param percent 歌曲加载进度
     * @param status 歌曲加载的状态
     * @param msg
     * @param lyricUrl
     */
    fun onMusicLoadProgress(songCode: Long, percent: Int, status: MusicLoadStatus, msg: String?, lyricUrl: String?)
}

/**
 * 切换演唱角色状态接口
 */
interface OnSwitchRoleStateListener {
    /**
     * 切换演唱角色成功
     */
    fun onSwitchRoleSuccess()

    /**
     * 切换演唱角色失败
     * @param reason 切换演唱角色失败的原因
     */
    fun onSwitchRoleFail(reason: SwitchRoleFailReason)
}

/**
 * KTVApi事件回调
 */
abstract class IKTVApiEventHandler {
    /**
     * 播放器状态变化
     * @param state MediaPlayer 播放状态
     * @param error MediaPlayer Error 信息
     * @param isLocal 本地还是主唱端的 Player 信息
     */
    open fun onMusicPlayerStateChanged(
        state: Constants.MediaPlayerState, error: Constants.MediaPlayerError, isLocal: Boolean
    ) {
    }

    /**
     * 歌曲播放结束分数
     * @param score 歌曲结束的总分
     */
    open fun onSingingScoreResult(score: Float) {}

    /**
     * ktvApi内部角色切换
     * @param oldRole 老角色
     * @param newRole 新角色
     */
    open fun onSingerRoleChanged(oldRole: KTVSingRole, newRole: KTVSingRole) {}

    open fun onChorusChannelTokenPrivilegeWillExpire(token: String?) {}
}

/**
 * 初始化KTVApi的配置
 * @param appId 用来初始化 Mcc Engine
 * @param rtmToken 创建 Mcc Engine 需要
 * @param engine RTC engine 对象
 * @param channelName 频道号，子频道名以基于主频道名 + "_ex" 固定规则生成频道号
 * Player 状态同步
 * 打分结果同步
 * 歌词同步
 * pitch同步
 * 建议你为KTVApi单独创建一个新的dataStreamId
 * @param localUid 创建 Mcc engine 和 加入子频道需要用到
 * @param chorusChannelName 子频道名 加入子频道需要用到
 * @param chorusChannelToken 子频道token 加入子频道需要用到
 */
data class KTVApiConfig(
    val appId: String,
    val rtmToken: String,
    val engine: RtcEngine,
    val channelName: String,
    val localUid: Int,
    val chorusChannelName: String,
    val chorusChannelToken: String
)

/**
 * 加载歌曲的配置，不允许在一首歌没有load完成前（成功/失败均算完成）进行下一首歌的加载
 * @param autoPlay 是否自动播放歌曲（通常主唱选择true）默认为false
 * @param mode 歌曲加载的模式， 默认为音乐和歌词均加载
 * @param songCode 歌曲 id
 * @param mainSingerUid 主唱的 Uid，如果是伴唱，伴唱需要根据这个信息 mute 主频道主唱的声音
 */
data class KTVLoadMusicConfiguration(
    val autoPlay: Boolean = false,
    val songCode: Long,
    val mainSingerUid: Int,
    val mode: KTVLoadMusicMode = KTVLoadMusicMode.LOAD_MUSIC_AND_LRC
)

interface KTVApi {
    /**
     * 初始化内部变量/缓存数据，并注册相应的监听，必须在其他KTVApi调用前调用initialize初始化KTVApi
     * @param config
     */
    fun initialize(config: KTVApiConfig)

    /**
     * 订阅KTVApi事件, 支持多注册
     * @param ktvApiEventHandler KTVApi事件接口实例
     */
    fun addEventHandler(ktvApiEventHandler: IKTVApiEventHandler)

    /**
     * 取消订阅KTVApi事件
     * @param ktvApiEventHandler KTVApi事件接口实例
     */
    fun removeEventHandler(ktvApiEventHandler: IKTVApiEventHandler)

    /**
     * 清空内部变量/缓存，取消在initWithRtcEngine时的监听，以及取消网络请求等
     */
    fun release()

    /**
     * 获取歌曲榜单
     * @param onMusicChartResultListener 榜单列表回调
     */
    fun fetchMusicCharts(
        onMusicChartResultListener: (
            requestId: String?, // TODO 不需要？
            status: Int,        // status=2 时token过期
            list: Array<out MusicChartInfo>?
        ) -> Unit
    )

    /**
     * 根据歌曲榜单类型获取歌单
     * @param musicChartId 榜单id
     * @param page 歌曲列表回调
     * @param pageSize 歌曲列表回调
     * @param jsonOption 自定义过滤模式
     * @param onMusicCollectionResultListener 歌曲列表回调
     */
    fun searchMusicByMusicChartId(
        musicChartId: Int,
        page: Int,
        pageSize: Int,
        jsonOption: String,
        onMusicCollectionResultListener: (
            requestId: String?,  // TODO 不需要？
            status: Int,         // status=2 时token过期
            page: Int,
            pageSize: Int,
            total: Int,
            list: Array<out Music>?
        ) -> Unit
    )

    /**
     * 根据关键字搜索歌曲
     * @param keyword 关键字
     * @param page 歌曲列表回调
     * @param jsonOption 自定义过滤模式
     * @param onMusicCollectionResultListener 歌曲列表回调
     */
    fun searchMusicByKeyword(
        keyword: String,
        page: Int, pageSize: Int,
        jsonOption: String,
        onMusicCollectionResultListener: (
            requestId: String?,  // TODO 不需要？
            status: Int,         // status=2 时token过期
            page: Int,
            pageSize: Int,
            total: Int,
            list: Array<out Music>?
        ) -> Unit
    )

    /**
     * 异步加载歌曲，同时只能为一首歌loadSong，loadSong结果会通过回调通知业务层
     * @param config 加载歌曲配置
     * @param onMusicLoadStateListener 加载歌曲结果回调
     *
     * 推荐调用：
     * 歌曲开始时：
     * 主唱 loadMusic(KTVLoadMusicConfiguration(autoPlay=true, mode=LOAD_MUSIC_AND_LRC, songCode, mainSingerUid)) switchSingerRole(SoloSinger)
     * 观众 loadMusic(KTVLoadMusicConfiguration(autoPlay=false, mode=LOAD_LRC_ONLY, songCode, mainSingerUid))
     * 加入合唱时：
     * 准备加入合唱者：loadMusic(KTVLoadMusicConfiguration(autoPlay=false, mode=LOAD_MUSIC_ONLY, songCode, mainSingerUid))
     * loadMusic成功后switchSingerRole(CoSinger)
     */
    fun loadMusic(
        config: KTVLoadMusicConfiguration,
        onMusicLoadStateListener: OnMusicLoadStateListener
    )

    /**
     * 异步切换演唱身份，结果会通过回调通知业务层
     * @param newRole 新演唱身份
     * @param onSwitchRoleState 切换演唱身份结果
     *
     * 允许的调用路径：
     * 1、Audience -》SoloSinger 自己点的歌播放时
     * 2、Audience -》LeadSinger 自己点的歌播放时， 且歌曲开始时就有合唱者加入
     * 3、SoloSinger -》Audience 独唱结束时
     * 4、Audience -》CoSinger   加入合唱时
     * 5、CoSinger -》Audience   退出合唱时
     * 6、SoloSinger -》LeadSinger 当前第一个合唱者加入合唱时，主唱由独唱切换成领唱
     * 7、LeadSinger -》SoloSinger 最后一个合唱者退出合唱时，主唱由领唱切换成独唱
     * 8、LeadSinger -》Audience 以领唱的身份结束歌曲时
     */
    fun switchSingerRole(
        newRole: KTVSingRole,
        onSwitchRoleStateListener: OnSwitchRoleStateListener?
    )

    /**
     * 播放歌曲
     * @param startPos 开始播放的位置
     * 对于主唱：
     * 如果loadMusic时你选择了autoPlay = true 则不需要主动调用startSing
     * 如果loadMusic时你选择了autoPlay = false 则需要在loadMusic成功后调用startSing
     * TODO 如何支持URL
     */
    fun startSing(songCode: Long, startPos: Long)

    /**
     * 恢复播放
     */
    fun resumeSing()

    /**
     * 暂停播放
     */
    fun pauseSing()

    /**
     * 调整进度
     */
    fun seekSing(time: Long)

    /**
     * 设置歌词组件，在任意时机设置都可以生效
     * @param view 传入的歌词组件view， 需要继承ILrcView并实现ILrcView的三个接口
     */
    fun setLrcView(view: ILrcView)

    /**
     * 设置当前mic开关状态
     * 目前关麦调用 adjustRecordSignalVolume(0) 后 onAudioVolumeIndication 仍然会执行， ktvApi需要增加一个变量判断当前是否关麦， 如果关麦把设置给歌词组件的pitch改为0
     */
    fun setMicStatus(isOnMicOpen: Boolean)

    /**
     * 获取mpk实例
     */
    fun getMediaPlayer() : IMediaPlayer

    /**
     * 获取mcc实例
     */
    fun getMusicCenter() : IAgoraMusicContentCenter
}

// 0、注释换成Print 不要有AgoraPrint
// 1、token 和 2ndChannelName 传递方式和业务逻辑调用
// 2、主动暴露 getMusicCenter() 为了renewToken
// 3、onChorusChannelTokenPrivilegeWillExpire

File: iOS/AgoraEntScenarios/Scenes/KTV/KTVAPI/KTVApi.swift
//
//  KTVApiDelegate.swift
//  AgoraEntScenarios
//
//  Created by wushengtao on 2023/3/14.
//

import Foundation
import AgoraRtcKit

/// 用户角色
@objc public enum KTVSingRole: Int {
    case soloSinger = 0     //独唱者
    case coSinger           //伴唱
    case leadSinger         //主唱
    case audience           //观众
//    case followSinger       //跟唱
}

@objc public enum loadMusicType: Int {
    case mcc
    case local
}

/// 歌曲状态
@objc public enum KTVPlayerTrackMode: Int {
    case origin = 0    //原唱
    case lead          //导唱
    case acc           //伴奏
}

/// 加载歌曲失败原因
@objc public enum KTVLoadMusicMode: Int {
    case loadNone
    case loadMusicOnly
    case loadLrcOnly
    case loadMusicAndLrc
}

/// 加载歌曲失败原因
@objc public enum KTVLoadSongFailReason: Int {
    case noLyricUrl = 0         //无歌词
    case musicPreloadFail   //歌曲预加载失败
    case cancled // 本次加载取消/停止
}

@objc public enum KTVSwitchRoleState: Int {
    case success = 0
    case fail
}

@objc public enum KTVSwitchRoleFailReason: Int {
    case none = 0
    case joinChannelFail
    case noPermission
}


/// 加入合唱结果状态
@objc public enum KTVJoinChorusState: Int {
    case success = 0    //加入合唱成功
    case fail           //加入合唱失败
}


/// 加入合唱失败原因
@objc public enum KTVJoinChorusFailReason: Int {
    case musicPreloadFail  //歌曲预加载失败
    case musicOpenFail     //歌曲打开失败
    case joinChannelFail   //加入ex频道失败
    case musicPreloadFailAndJoinChannelFail
}

@objc public enum KTVType: Int {
    case normal
    case singbattle
}

@objc public protocol IMusicLoadStateListener: NSObjectProtocol {
    
    
    /// 歌曲进度
    /// - Parameters:
    ///   - songCode: <#songCode description#>
    ///   - percent: 歌曲加载进度 范围： 0-100
    ///   - status: <#status description#>
    ///   - msg: <#msg description#>
    ///   - lyricUrl: <#lyricUrl description#>
    func onMusicLoadProgress(songCode: Int, percent: Int, status: AgoraMusicContentCenterPreloadStatus, msg: String?, lyricUrl: String?)
    
    /// 歌曲加载成功
    /// - Parameters:
    ///   - songCode: 歌曲code
    ///   - lyricUrl: 歌词远端url
    func onMusicLoadSuccess(songCode: Int, lyricUrl: String)

    
    /// 歌曲加载失败
    /// - Parameters:
    ///   - songCode: 歌曲code
    ///   - lyricUrl: 歌曲远端url
    ///   - reason: 错误原因
    func onMusicLoadFail(songCode: Int, reason: KTVLoadSongFailReason)
}


//public protocol KTVJoinChorusStateListener: NSObjectProtocol {
//
//    /// 加入合唱成功
//    func onJoinChorusSuccess()
//
//    /// 加入合唱失败
//    /// - Parameter reason: 失败原因
//    func onJoinChorusFail(reason: KTVJoinChorusFailReason)
//}

@objc public protocol KTVLrcViewDelegate: NSObjectProtocol {
    func onUpdatePitch(pitch: Float)
    func onUpdateProgress(progress: Int)
    func onDownloadLrcData(url: String)
    func onHighPartTime(highStartTime: Int, highEndTime: Int)
}

@objc public protocol KTVApiEventHandlerDelegate: NSObjectProtocol {
    
    /// 歌曲播放状态变化
    /// - Parameters:
    ///   - state: <#state description#>
    ///   - error: <#error description#>
    ///   - isLocal: <#isLocal description#>
    func onMusicPlayerStateChanged(state: AgoraMediaPlayerState,
                                   error: AgoraMediaPlayerError,
                                   isLocal: Bool)
    
    
    /// 歌曲得分回调
    /// - Parameter score: <#score description#>
    func onSingingScoreResult(score: Float)
     
    /// 角色切换回调
    /// - Parameters:
    ///   - oldRole: <#oldRole description#>
    ///   - newRole: <#newRole description#>
    func onSingerRoleChanged(oldRole: KTVSingRole, newRole: KTVSingRole)
    
    func onTokenPrivilegeWillExpire()
    
    /**
         * 合唱频道人声音量提示
         * @param speakers 不同用户音量信息
         * @param totalVolume 总音量
         */
    func onChorusChannelAudioVolumeIndication(
        speakers: [AgoraRtcAudioVolumeInfo],
        totalVolume: Int)
}

@objc open class KTVApiConfig: NSObject{
    var appId: String
    var rtmToken: String
    weak var engine: AgoraRtcEngineKit?
    var channelName: String
    var localUid: Int = 0
    var chorusChannelName: String
    var chorusChannelToken: String
    var type: KTVType = .normal
    var maxCacheSize: Int = 10
    var musicType: loadMusicType = .mcc
    var isDebugMode: Bool = false
    @objc public
    init(appId: String,
         rtmToken: String,
         engine: AgoraRtcEngineKit,
         channelName: String,
         localUid: Int,
         chorusChannelName: String,
         chorusChannelToken: String,
         type: KTVType,
         maxCacheSize: Int,
         musicType: loadMusicType,
         isDebugMode: Bool
    ) {
        self.appId = appId
        self.rtmToken = rtmToken
        self.engine = engine
        self.channelName = channelName
        self.localUid = localUid
        self.chorusChannelName = chorusChannelName
        self.chorusChannelToken = chorusChannelToken
        self.type = type
        self.maxCacheSize = maxCacheSize
        self.musicType = musicType
        self.isDebugMode = isDebugMode
    }
}

/// 歌曲加载配置信息
@objcMembers open class KTVSongConfiguration: NSObject {
    public var songIdentifier: String = ""
    public var autoPlay: Bool = false   //是否加载完成自动播放
    public var mainSingerUid: Int = 0     //主唱uid
    public var mode: KTVLoadMusicMode = .loadMusicAndLrc
    
    func printObjectContent() -> String {
        var content = ""
        
        let mirror = Mirror(reflecting: self)
        for child in mirror.children {
            if let propertyName = child.label {
                if let propertyValue = child.value as? CustomStringConvertible {
                    content += "\(propertyName): \(propertyValue)\n"
                } else {
                    content += "\(propertyName): \(child.value)\n"
                }
            }
        }
        
        return content
   }
}


public typealias LyricCallback = ((String?) -> Void)
public typealias LoadMusicCallback = ((AgoraMusicContentCenterPreloadStatus, NSInteger) -> Void)
public typealias ISwitchRoleStateListener = (KTVSwitchRoleState, KTVSwitchRoleFailReason) -> Void
public typealias MusicChartCallBacks = (String, AgoraMusicContentCenterStatusCode, [AgoraMusicChartInfo]?) -> Void
public typealias MusicResultCallBacks = (String, AgoraMusicContentCenterStatusCode, AgoraMusicCollection) -> Void
public typealias JoinExChannelCallBack = ((Bool, KTVJoinChorusFailReason?)-> Void)

@objc public protocol KTVApiDelegate: NSObjectProtocol {
    
    /// 初始化
    /// - Parameter config: <#config description#>
    init(config: KTVApiConfig)
    
    
    /// 订阅KTVApi事件
    /// - Parameter ktvApiEventHandler: <#ktvApiEventHandler description#>
    func addEventHandler(ktvApiEventHandler: KTVApiEventHandlerDelegate)
    
    
    /// 取消订阅KTVApi事件
    /// - Parameter ktvApiEventHandler: <#ktvApiEventHandler description#>
    func removeEventHandler(ktvApiEventHandler: KTVApiEventHandlerDelegate)
    
    
    /// 清空内部变量/缓存，取消在initWithRtcEngine时的监听，以及取消网络请求等
    func cleanCache()
    
    /**
     * 收到 IKTVApiEventHandler.onTokenPrivilegeWillExpire 回调时需要主动调用方法更新Token
     * @param rtmToken musicContentCenter模块需要的rtm token
     * @param chorusChannelRtcToken 合唱需要的频道rtc token
     */
    func renewToken(
        rtmToken: String,
        chorusChannelRtcToken: String)
    
    /**
     * 获取歌曲榜单
     * Parameter completion: 榜单列表回调
     */
    func fetchMusicCharts(completion:@escaping MusicChartCallBacks)
    
    /**
    * 根据歌曲榜单类型搜索歌单
    *  Parameters:
    *  musicChartId: 榜单id
    *  page: 榜单的查询页数
    *  pageSize: 查询每页的数据长度
    *  jsonOption: 自定义过滤模式
    *  completion: 歌曲列表回调
    */
    func searchMusic(musicChartId: Int,
                     page: Int,
                     pageSize: Int,
                     jsonOption: String,
                     completion:@escaping MusicResultCallBacks)
    
    /**
    * 根据关键字搜索歌曲
    *  Parameters:
    *  keyword: 搜索关键字
    *  page: 榜单的查询页数
    *  pageSize: 查询每页的数据长度
    *  jsonOption: 自定义过滤模式
    *  completion: 歌曲列表回调
    */
    func searchMusic(keyword: String,
                     page: Int, pageSize: Int,
                     jsonOption: String,
                     completion: @escaping MusicResultCallBacks)
            
    
    
    /// 加载歌曲
    /// - Parameters:
    ///   - config: <#config description#>
    ///   - onMusicLoadStateListener: <#onMusicLoadStateListener description#>
    func loadMusic(songCode: Int, config: KTVSongConfiguration, onMusicLoadStateListener: IMusicLoadStateListener)
    
    /// 通过url加载歌曲
    /// - Parameters:
    ///   - config: <#config description#>
    ///   - onMusicLoadStateListener: <#onMusicLoadStateListener description#>
    func loadMusic(config: KTVSongConfiguration, url: String)
    
    
    /// 切换角色
    /// - Parameters:
    ///   - newRole: <#newRole description#>
    ///   - token: <#token description#>
    ///   - onSwitchRoleState: <#onSwitchRoleState description#>
    func switchSingerRole(newRole: KTVSingRole, onSwitchRoleState:@escaping ISwitchRoleStateListener)
    
    
    /// 播放
    /// - Parameter startPos: <#startPos description#>
    func startSing(songCode: Int, startPos: Int)
    
    /**
     * 播放歌曲
     * @param url 歌曲地址
     * @param startPos 开始播放的位置
     * 对于主唱：
     * 如果loadMusic时你选择了autoPlay = true 则不需要主动调用startSing
     * 如果loadMusic时你选择了autoPlay = false 则需要在loadMusic成功后调用startSing
     */
    func startSing(url: String, startPos: Int)
    
    /// 恢复播放
    func resumeSing()
    
    
    /// 暂停播放
    func pauseSing()
    
    
    /// 调整进度
    /// - Parameter time: 进度，单位ms
    func seekSing(time: Int)
    
    /**
     * 设置当前音频播放delay， 适用于音频自采集的情况
     * @param audioPlayoutDelay 音频帧处理和播放的时间差
     */
    func setAudioPlayoutDelay(audioPlayoutDelay: Int)
    
    /// 设置歌词组件，在任意时机设置都可以生效
    /// - Parameter view: <#view description#>
    func setLrcView(view: KTVLrcViewDelegate)
    
    
    /// 设置当前mic开关状态目前关麦调用
    /// 目前关麦调用 adjustRecordSignalVolume(0) 后 onAudioVolumeIndication 仍然会执行， ktvApi需要增加一个变量判断当前是否关麦， 如果关麦把设置给歌词组件的pitch改为0
    /// - Parameter isOnMicOpen: <#isOnMicOpen description#>
    func setMicStatus(isOnMicOpen: Bool)
    
    func getMusicPlayer() -> AgoraRtcMediaPlayerProtocol?
    
    /// 获取MCC实例
    /// - Returns: <#description#>
    func getMusicContentCenter() -> AgoraMusicContentCenter?
    
    // 开启专业主播模式
    func enableProfessionalStreamerMode(_ enable: Bool)
    
    /**
     创建dataStreamID
     */
    func renewInnerDataStreamId()
    
    
  /**
   * 加载歌曲，同时只能为一首歌loadSong，同步调用， 一般使用此loadSong是歌曲已经preload成功（url为本地文件地址）
   * @param config 加载歌曲配置，config.autoPlay = true，默认播放url1
   * @param url1 歌曲地址1
   * @param url2 歌曲地址2
   *
   *
   * 推荐调用：
   * 歌曲开始时：
   * 主唱 loadMusic(KTVSongConfiguration(autoPlay=true, mode=LOAD_MUSIC_AND_LRC, url, mainSingerUid)) switchSingerRole(SoloSinger)
   * 观众 loadMusic(KTVSongConfiguration(autoPlay=false, mode=LOAD_LRC_ONLY, url, mainSingerUid))
   * 加入合唱时：
   * 准备加入合唱者：loadMusic(KTVSongConfiguration(autoPlay=false, mode=LOAD_MUSIC_ONLY, url, mainSingerUid))
   * loadMusic成功后switchSingerRole(CoSinger)
   */
  func load2Music(
      url1: String,
      url2: String,
      config: KTVSongConfiguration
  )
  
  /**
   * 多文件切换播放资源
   * @param url 需要切换的播放资源，需要为 load2Music 中 参数 url1，url2 中的一个
   * @param syncPts 是否同步切换前后的起始播放位置: true 同步，false 不同步，从 0 开始
   */
  func switchPlaySrc(url: String, syncPts: Bool)
      
}
